sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala

```diff
@@ -47,20 +49,49 @@ object ResolveHints {
    *
    * This rule must happen before common table expressions.
    */
-  class ResolveBroadcastHints(conf: SQLConf) extends Rule[LogicalPlan] {
+  class ResolveBroadcastHints(conf: SQLConf, catalog: SessionCatalog) extends Rule[LogicalPlan] {
     private val BROADCAST_HINT_NAMES = Set("BROADCAST", "BROADCASTJOIN", "MAPJOIN")
 
     def resolver: Resolver = conf.resolver
 
-    private def applyBroadcastHint(plan: LogicalPlan, toBroadcast: Set[String]): LogicalPlan = {
+    private def namePartsWithDatabase(nameParts: Seq[String], database: String): Seq[String] = {
+      if (nameParts.size == 1) {
+        database +: nameParts
+      } else {
+        nameParts
+      }
+    }
+
+    private def matchedTableIdentifier(
+        nameParts: Seq[String],
+        tableIdent: IdentifierWithDatabase): Boolean = {
+      tableIdent.database match {
+        case Some(db) if resolver(catalog.globalTempViewManager.database, db) =>
+          val identifierList = db :: tableIdent.identifier :: Nil
+          namePartsWithDatabase(nameParts, catalog.globalTempViewManager.database)
+            .corresponds(identifierList)(resolver)
+        case None if catalog.getTempView(tableIdent.identifier).isDefined =>
+          nameParts.size == 1 && resolver(nameParts.head, tableIdent.identifier)
+        case _ =>
+          val db = tableIdent.database.getOrElse(catalog.getCurrentDatabase)
```

Comment:
do we really need to support `spark.table("t").hint("default.t")`?

Reply:
I don't have a strong opinion on this though, better to remove this?
For example, in that case, it just ignores hints and prints warnings? https://issues.apache.org/jira/browse/SPARK-26820

Reply:
We should clearly define the rule of hint resolution first. For example
```
SELECT /* BROADCAST(t) */ * FROM db1.t JOIN db2.t 
```
Which table we should apply the broadcast hint?

Reply:
Sorry, but I forgot the previous discussion: https://github.com/apache/spark/pull/22198#issuecomment-416667343

To avoid the complexity, we would better to keep the current behaviour: it just ignores a database name in multi-part names. So, in the example you described, we assume Spark simply  applies the hint into both. Thought?

I don't update this pr yet cuz I just was waiting for other developer's comments.

Reply:
Then let's keep it simple and ignore the current database. We should only match a table iff
1. table name matches if the hint table name only has one part.
2. table name and qualifiers both match, if the hint table name has more than one name part.

This means, `SELECT /* BROADCAST(t) */ * FROM db1.t JOIN db2.t` will match both tables, and `SELECT /* BROADCAST(default.t) */ * FROM t` match no table.

Reply:
ok, thanks for the check. I'll update based on the comment.